nemo/collections/nlp/modules/common/text_generation_strategy.py

```diff
@@ -176,6 +180,65 @@ def post_generation_process(self, output):
         """
         return output
 
+    def _get_end_of_generation_tokens_and_strings(
+        self, eod_id: int, end_strings: List[str]
+    ) -> Tuple[Set[int], List[str]]:
+        """
+        return the tokens and strings indicating the end of generation
+        Args:
+            eod_id (int): the end of document token id
+            end_strings (List[str]): the list of end of generation strings
+        Returns:
+            a pair `(tokens, strings)` where `tokens` is a set of tokens (int) and `strings` is a list of strings,
+            which must all be used to identify the end of generation (`tokens` always contains `eod_id`, while
+            `strings` may be empty if all end strings are associated to unique tokens)
+        """
+        tokenizer = self.model.tokenizer
+        # A cache is used to remember which end strings are associated to unique tokens vs. which ones
+        # require an actual string comparison.
+        if self._end_of_generation_cache is None or self._end_of_generation_cache["tokenizer"] is not tokenizer:
+            # Invalidate the cache.
+            self._end_of_generation_cache = {
+                "tokenizer": tokenizer,
+                "end_string_to_token": {END_OF_SEQ: eod_id},
+                "end_strings_to_check": set(),
+            }
+        end_string_to_token = self._end_of_generation_cache["end_string_to_token"]
+
+        end_tokens = {eod_id}  # always include `eod_id`, even if `END_OF_SEQ` is not within `end_strings`
+        end_strings_to_check = []  # will contain end strings that have no associated special token
+
+        for end_string in end_strings:
+            try:
+                end_tokens.add(end_string_to_token[end_string])
+                continue
+            except KeyError:
+                if end_string in self._end_of_generation_cache["end_strings_to_check"]:
+                    end_strings_to_check.append(end_string)
+                    continue
+
+            # `end_string` does not exist in the cache yet: check if `end_string` is a special token for
+            # the tokenizer. Ideally, we would simply use `tokenizer.text_to_ids(end_string)`, but some
+            # tokenizers (e.g., SentencePiece) may prefix the special token with another token associated
+            # to an empty string. The code below is thus meant to extract the special token associated to
+            # `end_string` (if it exists). Note that using "This is a sequence." as reference is arbitrary.
+            ids_ref = tokenizer.text_to_ids("This is a sequence.")
```

Comment:
`"This is a sequence."` is not a good prefix string since the `.` character might be merged with other characters in the `end_string`.  Ideally we should use some special token so it will always stand alone.

Reply:
was exactly commenting this, but the code checks `ids_with_end_string[:-1] == ids_ref` below and also `.` is usually an independent token in most tokenizers

ideally a seaprate special token might be better, but do we have any token that we can guarantee will always be present?

Reply:
odds wise, I feel having `.` as the special token is more likely than any other thing. worst case we do a string check but it still doesnt give incorrect answer

Reply:
I started to write an explanation for my reasoning here, but this brought up a doubt in my mind that is very important to clear up, as it may invalidate a strong assumption made in this PR:

Is it possible that we may want to use an end string associated to a unique token (e.g., `<extra_id_1>`) and yet expect the model to end a response with the *string* `"<extra_id_1>"` but *without* generating this token?
I assumed no (similar to how we don't stop generation if the model generates the string "<|endoftext|>" instead of `eos_id`), but can you confirm this is correct?

Reply:
> `"This is a sequence."` is not a good prefix string since the `.` character might be merged with other characters in the `end_string`. Ideally we should use some special token so it will always stand alone.

So, following offline discussion, there are three cases:

1. `end_string` is actually a special token for the tokenizer (e.g. `<extra_id_1>` typically is, when we use it). Then we are guaranteed that it will be tokenized as a single token, and we will identify it properly.

2. `end_string` is *not* a special token for the tokenizer, and is tokenized into more than one token: in that case we don't care whether or not it is merged with the preceding `.`, because we will need to use string comparisons anyway.

3. `end_string` is *not* a special token for the tokenizer, and is tokenized into a single token: in that case either we identify the single token and rely on token comparison, or we don't (because of the tokenizer merging the `.`) and we fall back to string matching. I would argue that the latter is safer, because (a) it will ensure we always end generation correctly, and (b) it shows a warning that may alert the user that something may be off (there's a good chance they didn't expect the tokenizer to merge `end_string` with other characters). And if the tokenizer merges the `.` with `end_string` it seems particularly important to be aware of it, since this sequence of characters is likely to be quite common to finish responses. So it seems to me that it's actually a better option than `<extra_id_1>` (though I admit I hadn't thought it through before ;))

Reply:
the assmption is either end_string is a special token or not, the generation will end with it inclusively. It is up to tokenizer to decide whether the `ids_to_text` method wants to show it. I think `<extra_id_1>` is a better prefix because if it is a special token, it will work as expected. If it is not, the `>` is less likely to be merged with the end_string, which is a cleaner case as you mentioned in the 3.  If '.' is merged with the other partial end-string during the generation, the string match might not capture it correctly. E.g. end_string is "hello", '.' is merged to 'hel', the generated tokens are '.hel', 'lo,'  it won't trigger the string match check as we use 'endswith` method.

Reply:
Alright, let's go with `<extra_id_1>`, I don't think it's a big deal anyway, it shouldn't matter much in practice. This is done in c9a6d71f2778b7b85407e8aeca3d3f3e5b6df55e (I also rebased on top of main, but there are no changes in previous commits).

I still want to address the point below since a similar situation may still happen with `<extra_id_1>`:

> If '.' is merged with the other partial end-string during the generation, the string match might not capture it correctly. E.g. end_string is "hello", '.' is merged to 'hel', the generated tokens are '.hel', 'lo,' it won't trigger the string match check as we use 'endswith` method.

If '.' is merged with 'hel', then this will trigger the string match because the second condition of this check won't be satisfied (we  check both that there are N+1 tokens *and* also that the first N tokens are the same):
```python
if len(ids_with_end_string) == len(ids_ref) + 1 and ids_with_end_string[:-1] == ids_ref:
```
As a result, there will be a warning displayed, and the comparison will be made with `text.endswith("hello")` which will match any generation ending with "hello", regardless of what tokens this corresponds to.
Note that there could still be situations where the model generates "hello" without generation ending, e.g. if it generates tokens ".hel" followed by "lo world". But this case was not handled previously either, and it is unclear that we should stop there (since we can't truncate the model output mid-token, so the response would actually end with "hello world" rather than "hello".